src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,304 @@
+	Message.push_back('\0');
+	Message.push_back('\0');
+
+	m_Client->SendPluginMessage("FML|HS", Message);
+}
+
+AStringMap cForgeHandshake::ParseModList(const char * a_Data, size_t a_Size)
+{
+	cByteBuffer Buf(a_Size);
+	Buf.Write(a_Data, a_Size);
+
+	Int8 Discriminator;
+	Buf.ReadBEInt8(Discriminator);
+
+	ASSERT(Discriminator == 2);
+
+	UInt32 NumMods;
+	Buf.ReadVarInt32(NumMods);
+
+	AStringMap Mods;
+
+	for (size_t i = 0; i < NumMods; ++i)
+	{
+		AString Name, Version;
+		Buf.ReadVarUTF8String(Name);
+		Buf.ReadVarUTF8String(Version);
+
+		Mods.insert(std::pair<AString, AString>(Name, Version));
+
+		LOGD("ParseModList name=%s, version=%s", Name.c_str(), Version.c_str());
+	}
+
+	return Mods;
+}
+
+void cForgeHandshake::DataReceived(cClientHandle * a_Client, const char * a_Data, size_t a_Size)
+{
+	if (!m_IsForgeClient)
+	{
+		LOG("Received unexpected Forge data from non-Forge client (%zu bytes)", a_Size);
+		return;
+	}
+
+	LOGD("Received Forge data: %zu bytes: %s", a_Size, a_Data);
+
+	if (a_Size <= 1)
+	{
+		LOG("Received unexpectedly short Forge data (%zu bytes)", a_Size);
+		return;
+	}
+
+	int Discriminator = a_Data[0];
+
+	switch (Discriminator)
+	{
+		case Discriminator_ClientHello:
+		{
+			if (a_Size == 2)
+			{
+				int FmlProtocolVersion = a_Data[1];
+				LOGD("Received ClientHello with FML protocol version %d", FmlProtocolVersion);
+				if (FmlProtocolVersion != 2)
+				{
+					LOG("Unsupported FML client protocol version received in ClientHello: %d", FmlProtocolVersion);
+					SetError();
+				}
+			}
+			else
+			{
+				LOG("Unexpectedly short ClientHello received");
+				SetError();
+			}
+
+			break;
+		}
+
+		case Discriminator_ModList:
+		{
+			LOGD("Received ModList");
+
+			AStringMap ClientMods = ParseModList(a_Data, a_Size);
+			AString ClientModsString;
+			for (auto & item: ClientMods)
+			{
+				ClientModsString.append(item.first);
+				ClientModsString.append("@");
+				ClientModsString.append(item.second);
+				ClientModsString.append(", ");
+			}
+
+			LOG("Client connected with %zu mods: %s", ClientMods.size(), ClientModsString.c_str());
+
+			if (m_Client->m_ForgeMods != nullptr)
+			{
+				delete m_Client->m_ForgeMods;
+			}
+
+			m_Client->m_ForgeMods = new cForgeMods(ClientMods);
+
+			// Let the plugins know about this event, they may refuse the player:
+			if (cRoot::Get()->GetPluginManager()->CallHookLoginForge(*a_Client))
+			{
+				LOG("Modded client refused by plugin");
+				SetError();
+				return;
+			}
+
+			// Send server ModList
+
+			// TODO: max size?
+			cByteBuffer Buf(1024);
+
+			// Send server-side Forge mods registered by plugins
+			auto &ServerMods = m_Client->GetForgeMods();
+
+			size_t ModCount = ServerMods.GetNumMods();
+
+			Buf.WriteBEInt8(Discriminator_ModList);
+			Buf.WriteVarInt32(static_cast<UInt32>(ModCount));
+			for (size_t i = 0; i < ModCount; ++i)
+			{
+				Buf.WriteVarUTF8String(ServerMods.GetModNameAt(i));
+				Buf.WriteVarUTF8String(ServerMods.GetModVersionAt(i));
```

Comment:
cItems (in Item.h:223) is a class built solely to wrap around a vector of items. There they do it also so that they can raise appropriate warnings for invalid parameters.

So I don't know if tolua can bridge fancy types, but whoever wrote cItems didn't think it could (and I think it'd require the std::map class declaration being marked as tolua_export, or manual bindings).